pkg/vm/vm.go

```diff
@@ -624,14 +624,23 @@ func EncodeMemory(memory []*f.Element) []byte {
 }
 
 func DecodeMemory(content []byte) []*f.Element {
+	if len(content) == 0 {
+		return make([]*f.Element, 0)
+	}
+
 	// calculate the max memory index
-	lastContentInd := len(content) - (addrSize + feltSize)
-	lasMemIndex := binary.LittleEndian.Uint64(content[lastContentInd : lastContentInd+addrSize])
+	lastMemIndex := uint64(0)
+	for i := 0; i < len(content); i += addrSize + feltSize {
+		memIndex := binary.LittleEndian.Uint64(content[i : i+addrSize])
+		if memIndex > lastMemIndex {
+			lastMemIndex = memIndex
+		}
+	}
```

Comment:
Could you add a small comment explaining the reason why we are using the for to scan for the whole memory